presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateProjectOnValues.java

```diff
@@ -0,0 +1,142 @@
+import static com.facebook.presto.sql.planner.plan.Patterns.project;
+import static com.facebook.presto.sql.planner.plan.Patterns.source;
+import static com.facebook.presto.sql.planner.plan.Patterns.values;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static java.util.Objects.requireNonNull;
+import static org.glassfish.jersey.internal.util.collection.ImmutableCollectors.toImmutableList;
+
+/**
+ * This optimizer looks for ProjectNode followed by a ValuesNode and get the ProjectNode Evaluated.
+ * When this rule is used on iterative optimizer, the rule could apply iteratively.
+ * <p/>
+ * Plan before optimizer:
+ * <pre>
+ * ProjectNode (outputVariables)
+ *   - ValuesNode
+ * </pre>
+ * <p/>
+ * Plan after optimizer:
+ * <pre>
+ * ValuesNode (outputVariables)
+ * </pre>
+ */
+public class EvaluateProjectOnValues
+        implements Rule<ProjectNode>
+{
+    private static final Pattern<ProjectNode> PATTERN = project().with(source().matching(values()));
+
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public EvaluateProjectOnValues(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionManager is null");
+    }
+
+    @Override
+    public Pattern<ProjectNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isEvaluateProjectOnValues(session);
+    }
+
+    @Override
+    public Result apply(ProjectNode projectNode, Captures captures, Context context)
+    {
+        Optional<PlanNode> optionalSource = context.getLookup().resolveGroup(projectNode.getSource()).findFirst();
+        if (!optionalSource.isPresent()) {
+            return Result.empty();
+        }
+        ValuesNode source = (ValuesNode) optionalSource.get();
+        List<List<RowExpression>> rows = source.getRows();
+        List<VariableReferenceExpression> valuesOutputVariables = source.getOutputVariables();
+        Set<Map.Entry<VariableReferenceExpression, RowExpression>> projectAssignmentEntries = projectNode.getAssignments().entrySet();
+        List<VariableReferenceExpression> projectOutputVariables = projectAssignmentEntries.stream().map(Map.Entry::getKey).collect(toImmutableList());
+        List<RowExpression> projectRowExpressions = projectAssignmentEntries.stream().map(Map.Entry::getValue).collect(toImmutableList());
+        // exclude non-deterministic function
+        if (!projectRowExpressions.stream()
+                .filter(expression -> expression instanceof CallExpression)
+                .map(callExpression -> functionAndTypeManager.getFunctionMetadata(((CallExpression) callExpression).getFunctionHandle()))
+                .filter(functionMetadata -> !functionMetadata.isDeterministic()).collect(toImmutableList()).isEmpty()) {
```

Comment:
can be replaced with `.allMatch(FunctionMetadata::isDeterministic)`

Reply:
fixed.